Feature/minimal mode

crisner opened the pull request:
Fixes #123  (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

crisner commented:
Things done:
- created a toggle button
- Added toggle function
- Added circle marker and styles to eonet fires layer
- Added condition to render markers according to selected mode

Things to do:
- Add circle marker and conditions to remaining layers
- Add minimal mode control to one-liner code

### Default mode:
![127 0 0 1_5500_example_index html (11)](https://user-images.githubusercontent.com/29401459/72818618-c918ed80-3c91-11ea-9cd9-2abd0630bcae.png)

### Minimal mode:
![127 0 0 1_5500_example_index html (12)](https://user-images.githubusercontent.com/29401459/72818631-d03ffb80-3c91-11ea-8e02-63b05d13fc09.png)

The bottom-most button on the left side of the map is the button to toggle modes. The icon changes depending on the mode.

crisner commented:
The toggle button is a simple leaflet styled button for now. What I had in my project proposal was this:
![1](https://user-images.githubusercontent.com/29401459/72819040-83105980-3c92-11ea-8a5a-16cf8ca9067b.png)

This seemed a bit out of place so I opted for the simpler button. I could maybe make the button attached here to look more like leaflet styles if a switch type toggle button is preferred.

sagarpreet-chadha commented on example/layers.js:
```diff
@@ -264,6 +264,9 @@ var embedControl = new L.control.embed({
 });
 embedControl.addTo(map);
 
+var modeControl = new L.control.minimalMode(leafletControl);
```
So many features :P 
We need to make readme.md readble!!!

crisner replied:
Yes. I will be doing that this weel. :)

crisner commented:
### Minimal mode markers from aqicn and opensense layers:
![127 0 0 1_5500_example_index html (15)](https://user-images.githubusercontent.com/29401459/72894174-b90c1700-3d40-11ea-99ac-4ea9b95f0145.png)

In default mode the screen starts to lag when both these layers are turned on. The experience seems to be smooth in minimal mode.

crisner commented:
Also, the markers in aqicn layer are colour-coded and I carried the same over to minimal mode. So, when this layer is turned on with other layers it may be difficult to identify which layer the markers belong to. Should I change the markers to have a single colour?

crisner commented:
To do:
- add minimal mode to remaining layers
- update readme

nstjean commented:
The implementation looks good! Not sure about the button. I'm not sure people will have any idea what it does, though maybe the tooltip will be enough.

It looks like button shows the minimal dot when it's in minimal mode, and shows a map marker when the map markers are showing. Should that be opposite? So when you're in minimal mode it shows the full map marker on the button so you know to click it to show the full marker?

crisner commented:
You are right! It should be the opposite! I

More things to do:
- Correct the icon displayed on toggle button
- Add tooltip

I just got an idea. What if I add the toggle button like the zoom in and out buttons? One could be disabled while the other mode is active.

nstjean commented:
I think that would be nice and visual to see what is on/off!

crisner commented:
Active layers: aqicn, openaq and opensense
#### Default mode;
![127 0 0 1_5500_example_index html (17)](https://user-images.githubusercontent.com/29401459/72983809-39497f80-3e08-11ea-86d9-dab22293100c.png)
#### Minimal mode;
![127 0 0 1_5500_example_index html (16)](https://user-images.githubusercontent.com/29401459/72983784-28990980-3e08-11ea-8a8c-d3feacc81b71.png)

crisner commented:
To do:
- write tests

crisner commented:
Updated npm version by running `npm version minor`.

nstjean commented on src/util/modeControl.js:
```diff
@@ -0,0 +1,84 @@
+      this._map._minimalMode = this.options.minimalMode;
+      this._modeBtnContainer = L.DomUtil.create('div', 'leaflet-control-mode leaflet-bar leaflet-control');
+      this._defaultMode = this._createButton('Default mode', 'fas fa-map-marker-alt', this.loadDefaultMode);
+      this._minimalMode = this._createButton('Minimal mode', 'far fa-dot-circle', this.loadMinimalMode);
```
Maybe this could be something more descriptive, like "Default markers" and "Dot markers"? Or "Minimal markers"?

crisner replied:
Do you mean in L17 and L18? If so for clarity I could change it to defaultModeBtn? I think it will be misleading to add 'markers' to the button's name.

nstjean replied:
Yes, I mean for the text that is in the pop-up, so it is descriptive to users. Changing the name to `defaultModeBtn` would be good!

crisner replied:
Ah, that one! That's a great idea!

crisner commented:
Minimal mode in the demo page: https://crisner.github.io/leaflet-environmental-layers/example/index.html#lat=34.034&lon=-91.165&zoom=5&layers=Standard,fractracker
Minimal mode in the one-liner code: https://crisner.github.io/leaflet-environmental-layers/example/oneLinerCodeExample.html#lat=43.00&lon=-4.04&zoom=3&layers=Grey-scale,odorreport,eonetFiresLayer

jywarren commented:
Wow, really impressive!

![image](https://user-images.githubusercontent.com/24359/73280661-dca9e400-41bc-11ea-9e3b-5e9836b07bc5.png)

Ready for a merge?

jywarren commented:
These screenshots above would be great for a blog post!!! 😄

jywarren commented:
Ah, so, just needs a recompile. This is ready for merge otherwise? 🙌

crisner commented:
Yes. Ready for a merge. :)

jywarren commented:
Ah, sorry, i meant that it looks like the /dist/ files need to be recompiled, could you do that and then we'll merge? Thanks!

crisner commented:
- Merged and fixed conflicts
- Updated build
- Also added a demo page with multiple maps each having different options in the one-liner usage

![127 0 0 1_5500_example_multipleMaps html (1)](https://user-images.githubusercontent.com/29401459/73336723-9924a800-4298-11ea-9197-b33327fc8911.png)

@jywarren, this is now merge-ready. 😃

jywarren commented:
So awesome! Fantastic work, @crisner !!
